src/operator/contrib/multi_lamb-inl.h

```diff
@@ -0,0 +1,332 @@
+                            /multi_param->chunk_size;
+  }
+  memcpy(multi_param->step_count, p.step_count.begin(), multi_param->ntensors * sizeof(int));
+}
+
+using namespace mxnet_op;
+template<typename MPDType, typename DType>
+void call_kernel1(Stream<cpu>* s);
+template<typename MPDType, typename DType>
+void call_kernel1(Stream<gpu>* s);
+
+template<typename MPDType, typename DType>
+void call_kernel2(Stream<cpu>* s);
+template<typename MPDType, typename DType>
+void call_kernel2(Stream<gpu>* s);
+
+template<typename xpu, template<typename> class MPTypeChooser, int input_stride>
+inline void multiLAMB(const nnvm::NodeAttrs& attrs,
+                      const OpContext &ctx,
+                      const std::vector<TBlob> &inputs,
+                      const std::vector<OpReqType> &req,
+                      const std::vector<TBlob> &outputs) {
+  auto param = nnvm::get<MultiLAMBParam>(attrs.parsed);
+  Stream<xpu>* s = ctx.get_stream<xpu>();
+
+  MSHADOW_REAL_TYPE_SWITCH(inputs[0].type_flag_, DType, {
+    using MPDType = typename MPTypeChooser<DType>::type;
+    MultiLAMBKernelParam<DType, MPDType> kernel_params;
+    FillMultiLAMBKernelParam<xpu, DType, MPDType, MultiLAMBParam, input_stride>
+            (attrs, ctx, inputs, outputs, &kernel_params);
+
+    // create vector of TBlob with all the weights contiguous
+    std::vector<TBlob> weights;
+    for (size_t index = 0; index < kernel_params.ntensors; ++index) {
+        weights.emplace_back(inputs[index*input_stride]);
+    }
+    // create vector of TBlob with all the temp_g contiguous
+    std::vector<TBlob> temp_g;
+    for (size_t index = 0; index < kernel_params.ntensors; ++index) {
+        temp_g.emplace_back(inputs[index*input_stride+4]);
+    }
+
+    // Calculate amount of temporary storage
+    size_t workspace_size = 2 * kernel_params.ntensors * sizeof(float) +
+        2 * kernel_params.nchunks * sizeof(int);
+
+      // Request temporary storage
+    Tensor<xpu, 1, char> workspace =
+    ctx.requested[multilamb::kTempSpace].get_space_typed<xpu, 1, char>(
+      Shape1(workspace_size), s);
+
+    // Create tensors
+    size_t pos_wspace = 0;
+    Tensor<xpu, 1, float> r1(reinterpret_cast<float*>(&workspace[pos_wspace]),
+      Shape1(kernel_params.ntensors), s);
+    pos_wspace += kernel_params.ntensors * sizeof(float);
+    Tensor<xpu, 1, float> r2(reinterpret_cast<float*>(&workspace[pos_wspace]),
+      Shape1(kernel_params.ntensors), s);
+    pos_wspace += kernel_params.ntensors * sizeof(float);
+    Tensor<xpu, 1, int> block_to_tensor(reinterpret_cast<int*>(&workspace[pos_wspace]),
+      Shape1(kernel_params.nchunks), s);
+    pos_wspace += kernel_params.nchunks * sizeof(int);
+    Tensor<xpu, 1, int> block_to_chunk(reinterpret_cast<int*>(&workspace[pos_wspace]),
+      Shape1(kernel_params.nchunks), s);
+
+    MultiSumSqRun<xpu>(weights, kernel_params.ntensors, r1.dptr_, s);
+    call_kernel1<MPDType, DType>(s, kernel_params, param, block_to_tensor.dptr_,
```

Comment:
It looks like MultiSumSq uses atomic add and introduces non-determinism.. We should probably provide a way for users to have deterministic updates. I don't think we need a MultiLAMB optimizer, but rather we reuse the LAMB optimizer class and dispatch to the multi-lamb ops if aggregate size is larger than 1. 
This way, if a user wants deterministic updates, he/she can at least achieve that via setting the aggregation size env var.